problem1/problem1.cpp

```diff
@@ -0,0 +1,37 @@
+class Foo {
+private:
+    mutex m;
+    condition_variable cv;
+    int turn;
+    
+public:
+    Foo() : turn(0) { }
+
+    void first(function<void()> printFirst) {
+        turnWrapper(printFirst, 0);
+    }
+
+    void second(function<void()> printSecond) {
+        turnWrapper(printSecond, 1);
+    }
+
+    void third(function<void()> printThird) {
+        turnWrapper(printThird, 2);
+    }
+
+    void turnWrapper(function<void()> whichPrint, int whichTurn) {
+        unique_lock<mutex> lock(m);
+        cv.wait(lock, [this, whichTurn]{return turn == whichTurn;});
+
+        try {
+            whichPrint();
```

Comment:
oops, calling callback function (containing code you don't have control over) under the lock makes impossible to apply local reasoning to prove correctness of this code without looking at the whole program